### PR TITLE
Threshold

### PR DESCRIPTION
**Pull request by Henrod**
# Threshold

A new way to scale GRUs in Maestro

Watcher gets a window of time that is the scheduler's _autoscaling.up.trigger.time_. At each cycle (every _watcher.autoScalingPeriod_ seconds) if the usage is above _autoscaling.up.trigger.usage_, a counter is incremented. 

At the end of the window, if the counter over the number of points in this window is above _autoscaling.up.trigger.percentage_, the GRUs are scaled up. 

The same works for scale down.

### CODE REVIEW

File: manifests/scheduler-config-1.yaml [hunk 1]
```diff
@@ -16,9 +16,11 @@ autoscaling:
       usage: 70
       time: 10
     cooldown: 30
+      percentage: 80
```
Comment (felipejfc) [inline manifests/scheduler-config-1.yaml]: indentation seems wrong

File: manifests/scheduler-config-1.yaml [hunk 2]
```diff
@@ -16,9 +16,11 @@ autoscaling:
...
   down:
     delta: 1
     trigger:
       usage: 50
       time: 10
     cooldown: 30
+      percentage: 80
```
Comment (felipejfc) [inline manifests/scheduler-config-1.yaml]: here too

File: docs/api.md [hunk 3]
```diff
@@ -473,15 +475,17 @@ All API responses include a `X-Maestro-Version` header with the current Maestro
           "delta": 10,
           "trigger": {
             "usage": 70,
-            "time": 600
+            "time": 600,
```
Comment (felipejfc) [inline docs/api.md]: talvez seja melhor chamar de window agora

File: models/scheduler.go [hunk 4]
```diff
@@ -49,9 +49,13 @@ type EnvVar struct {
 }
 
 // ScalingPolicyTrigger has the configuration for a scaling policy trigger
+// During 'Time' seconds with n measures, 'Usage'% of the machines needs
+// to be occupied on 'Percentage'% of these n points.
+// This will trigger a scale up or scale down.
 type ScalingPolicyTrigger struct {
-	Time  int `yaml:"time" json:"time" valid:"int64"`
-	Usage int `yaml:"usage" json:"usage" valid:"int64"`
+	Time       int `yaml:"time" json:"time" valid:"int64"`
```
Comment (felipejfc) [inline models/scheduler.go]: Window

File: watcher/watcher.go [hunk 5]
```diff
@@ -50,6 +50,22 @@ type Watcher struct {
...
 	EventForwarders   []eventforwarder.EventForwarder
+	ScaleUpInfo       *ScaleInfo
+	ScaleDownInfo     *ScaleInfo
+}
+
+//ScaleInfo holds information about last time scheduler was verified if it needed to be scaled
+// and how many time it was above or below threshold
+type ScaleInfo struct {
+	LastTimeCheckedScale time.Time
+	PointsAboveThreshold int
```
Comment (felipejfc) [inline watcher/watcher.go]: PointsAboveOrBelowThreshold?

File: watcher/watcher.go [hunk 6]
```diff
@@ -375,54 +393,81 @@ func (w *Watcher) AutoScale() {
...
+		return false, false, changedState
 	}
 
 	if roomCount.Total() < autoScalingInfo.Min {
-		return true, false, false
+		return true, false, changedState
 	}
 
 	if 100*roomCount.Ready < (100-autoScalingInfo.Up.Trigger.Usage)*roomCount.Total() {
+		inSync = false
```
Comment (felipejfc) [inline watcher/watcher.go]: is this right? shouldn't we set inSync to false only when w.passedThreshold("up", autoScalingInfo)==true ?

File: watcher/watcher.go [hunk 7]
```diff
@@ -375,54 +393,81 @@ func (w *Watcher) AutoScale() {
...
 		}
-		inCooldownPeriod = true
+		w.ScaleUpInfo.LastTimeCheckedScale = now
+		w.ScaleUpInfo.PointsAboveThreshold = 0
 	}
 
+	downTriggerDuration := time.Duration(autoScalingInfo.Down.Trigger.Time) * time.Second
 	if 100*roomCount.Ready > (100-autoScalingInfo.Down.Trigger.Usage)*roomCount.Total() &&
 		roomCount.Total()-autoScalingInfo.Down.Delta >= autoScalingInfo.Min {
+		inSync = false
```
Comment (felipejfc) [inline watcher/watcher.go]: should we set inSync to false here?

File: watcher/watcher.go [hunk 8]
```diff
@@ -375,54 +393,81 @@ func (w *Watcher) AutoScale() {
...
-			return false, false, true
+			changedState = true
 		}
+	}
+	if w.ScaleDownInfo.LastTimeCheckedScale.Add(downTriggerDuration).Before(now) {
 		if nowTimestamp-scheduler.LastScaleOpAt > int64(autoScalingInfo.Down.Cooldown) &&
-			nowTimestamp-scheduler.StateLastChangedAt > int64(autoScalingInfo.Down.Trigger.Time) {
-			return false, true, false
+			w.passedThreshold("down", autoScalingInfo) {
+			shouldScaleDown = true
```
Comment (felipejfc) [inline watcher/watcher.go]: I guess that inSync should be set to false here

File: manifests/scheduler-config-1.yaml (hunk 1 above)
Comment (Henrod) [inline manifests/scheduler-config-1.yaml]: yes, ok

File: watcher/watcher.go (hunk 6 above)
Comment (Henrod) [inline watcher/watcher.go]: It's right. I want to indicate that the scheduler is over dimensioned before it scales down, for example. Setting here shows this log and when a cycle completes, it scales down, getting back to insync state.

File: watcher/watcher.go (hunk 5 above)
Comment (Henrod) [inline watcher/watcher.go]: ok

File: models/scheduler.go (hunk 4 above)
Comment (Henrod) [inline models/scheduler.go]: good, i will change that

File: controller/controller.go [hunk 9]
```diff
@@ -481,122 +502,147 @@ func UpdateSchedulerConfig(
...
+			if err != nil {
+				logger.WithError(err).Error("error when updating scheduler and waiting for new pods to run")
+				for _, podToDelete := range newPods {
+					deletePod(l, mr, clientset, redisClient.Client, configYAML.Name, podToDelete.GetName())
+					room := models.NewRoom(podToDelete.GetName(), schedulerName)
+					err := room.ClearAll(redisClient.Client)
 					if err != nil {
-						logger.WithError(err).Error("error when creating pod")
-						continue
+						//TODO: try again so Redis doesn't hold trash data
```
Comment (felipejfc) [inline controller/controller.go]: This TODO seems important

**Pull request closed without merging**